### ISSUE TITLE
Make newlines set cursor x position to 0

### ISSUE THREAD

**Issue by jwarkentin** [cronvel/terminal-kit#11]
When I output a chunk of text with newlines it looks very odd because the newline doesn't change the x position of the cursor. I don't know if there's a compelling reason not to move the cursor to the beginning of the line on a newline but if so it would be nice to either be able to set an option to do it or somehow manually do it. I would manually search and replace newlines with the right control codes in my output string if I knew what they were but it changes depending on the terminal and you've already built a system to figure all that out.

**Comment by jwarkentin** [comment]
Ok, so I don't know if it works for all terminals but if I replace all `\n`s with `\r\n` then it works for what I'm doing. It may still be a worthwhile option to consider adding but inserting the carriage return seems to get the job done for me right now.

**Comment by jwarkentin** [comment]
I think this same issue is manifesting itself in weird autocomplete menu and cursor placement. Here's an example:

![autocomplete menu](https://cloud.githubusercontent.com/assets/1084024/17046207/e7783ad2-4f8d-11e6-8a3e-8097c57d3129.png)

**Comment by cronvel** [comment]
That's strange... which OS and terminal are you using?
Single \n should works on all unixes.

Can you tell me if a simple `process.stdout.write( 'some\ntext' )` output `text` in the beginning of the line? 

I can only imagine this kind of things on Windows, and sadly, high-level features are not really compatible with Windows builtin-terminal (which terribly lacks of features).

**Comment by jwarkentin** [comment]
I'm probably using it different from the way most people use it. I'm using it in combination with the ssh2 package to provide a terminal over SSH from within my app. So I set the terminal-kit stdout to my SSH stream stdout and then I'm using the terminal over an actual SSH connection. That may affect this. Aside from that, I'm using it in gnome-terminal which reports over SSH as `xterm-256color`.

**Comment by cronvel** [comment]
Ok... Looks difficult to detect... Terminal-kit would think it runs inside a regular xterm-compatible terminal.

I don't know exactly what happend under the hood, *gnome-terminal* is probably set in a mode where it needs cr+lf, probably by SSH. Can you look at available SSH options and environment variable that are sent through SSH?

**Comment by jwarkentin** [comment]
I think I figured out what the problem is. Below I've posted all the info I receive from the SSH client initiating the connection. It sends the TTY mode settings it's requesting. I think the important one is `ONLRET` which makes newlines perform carriage returns. You can see what all of the modes mean [here](https://github.com/mscdex/ssh2#terminal-modes) or in RFC 4254 [here](https://www.ietf.org/rfc/rfc4254.txt).

However, it is the terminal's responsibility to do something with the mode settings sent and there's no way to pass those to terminal-kit as far as I can see. Also, that particular mode is not set in this case anyway so I'd have to override it. When I run `stty -a` in gnome-terminal it shows `olnret` is set so that is the normal behavior of a terminal.

When terminal-kit is run inside of a normal terminal provided by the system the outer terminal is handling that for you. Since I'm going over SSH and don't have the full terminal implementation surrounding terminal-kit the missing functionality becomes apparent.

So I guess this Github issue is my request for terminal-kit to implement a way to set at least some mode settings and implement the `ONLRET` setting functionality to perform a carriage return when a newline is encountered.

Don't know if this data will be helpful or not, but for the record here is the info the SSH client sends when it creates a new SSH session:

```
{
  "cols": 150,
  "rows": 40,
  "width": 0,
  "height": 0,
  "term": "xterm-256color",
  "modes": {
    "TTY_OP_OSPEED": 38400,
    "TTY_OP_ISPEED": 38400,
    "VINTR": 3,
    "VQUIT": 28,
    "VERASE": 127,
    "VKILL": 21,
    "VEOF": 4,
    "VEOL": 0,
    "VEOL2": 0,
    "VSTART": 17,
    "VSTOP": 19,
    "VSUSP": 26,
    "VREPRINT": 18,
    "VWERASE": 23,
    "VLNEXT": 22,
    "VDISCARD": 15,
    "IGNPAR": 0,
    "PARMRK": 0,
    "INPCK": 0,
    "ISTRIP": 0,
    "INLCR": 0,
    "IGNCR": 0,
    "ICRNL": 1,
    "IUCLC": 0,
    "IXON": 1,
    "IXANY": 0,
    "IXOFF": 0,
    "IMAXBEL": 0,
    "ISIG": 1,
    "ICANON": 1,
    "XCASE": 0,
    "ECHO": 1,
    "ECHOE": 1,
    "ECHOK": 1,
    "ECHONL": 0,
    "NOFLSH": 0,
    "TOSTOP": 0,
    "IEXTEN": 1,
    "ECHOCTL": 1,
    "ECHOKE": 1,
    "PENDIN": 0,
    "OPOST": 1,
    "OLCUC": 0,
    "ONLCR": 1,
    "OCRNL": 0,
    "ONOCR": 0,
    "ONLRET": 0,
    "CS7": 1,
    "CS8": 1,
    "PARENB": 0,
    "PARODD": 0
  }
}
```

**Comment by jwarkentin** [comment]
Don't know if this will be helpful but here is a snippet that effectively implements the `ONLRET` functionality if all strings sent to stdout and stderr use it:

```js
output.replace(/(\n+)/g, '\r$1')
```

**Comment by cronvel** [comment]
I'm still trying to figure out how to detect that.

- it looks like there is no escape sequences for line ending reporting
- Node.js does not provide an interface to modify TTY options (except setting raw mode on or off),
- spawning a program like `stty` is a really crappy solution, and would not work anyway through SSH

So we are in a dead-end... 

I have temporarly added an option for forcing `crlf` manually, in terminal-kit@0.24.16:

```js
var term = require( 'terminal-kit' ).terminal ;
term.options.crlf = true ;
term( 'Some\ntext\n' ) ;
```

If there is no way to detect, it will be documented and will becomes the *official* way of doing this.

Tell me if that solve all your issues...

**Comment by jwarkentin** [comment]
Thanks! That definitely solves the issue.

**Comment by jwarkentin** [comment]
In the case of SSH, terminal-kit is the only thing managing the input and output streams. There is no outter terminal to do any of this processing. It may be worth considering adding a way to actually pass in the input and output mode settings to terminal-kit and fully implementing them.

They could be passed in like this, based on my current implementation:
```js
let termkit = require('terminal-kit')

...

let term = this.term = termkit.createTerminal({
  stdin: stream.stdin,
  stdout: stream.stdout,
  stderr: stream.stderr,
  generic: this.ptyInfo.term,
  modes: this.ptyInfo.modes,    // Pass in the modes as shown in my comment above here
  appName: this.title
})
```

There is a lot of good information about implementing the modes [here](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap11.html#tag_11_02_03). One very relevant section states:

> If ONLCR is set, the NL character shall be transmitted as the CR-NL character pair. If OCRNL is set, the CR character shall be transmitted as the NL character. If ONOCR is set, no CR character shall be transmitted when at column 0 (first position). If ONLRET is set, the NL character is assumed to do the carriage-return function; the column pointer shall be set to 0 and the delays specified for CR shall be used. Otherwise, the NL character is assumed to do just the line-feed function; the column pointer remains unchanged. The column pointer shall also be set to 0 if the CR character is actually transmitted.

If you want, I'd even be willing to take a crack at implementing something like this in terminal-kit.

**Comment by cronvel** [comment]
I know those modes, because I have used them in C/C++ 10 or 15 years ago, but I just don't know how to get/set them in Node.js. If you know how to do it, go ahead! I would be glad to add that in terminal-kit!

**Comment by cronvel** [comment]
I got a response from Thomas Dickey from [invisible island](http://invisible-island.net/), the best resource for terminal: 

> [...]
> 
> One of my user have issues with line ending through SSH, it looks
> like a terminal that usually treats \n like line ending needs a \r\n
> through SSH.
> 
> So... is there an escape sequence asking the terminal in which mode
> it is running? 
> > no... the features known to 'stty' are done in the terminal driver,
> > (part of the operating system), not in the terminal itself.
> > 
> > You can get different results from 'stty', depending on whether there's
> > actually a terminal or not.  For instance, this is from Linux with
> > everything via pipes:
> > 
> > speed 38400 baud; rows 40; columns 80; line = 0;
> > intr = ^C; quit = ^\; erase = ^H; kill = ^U; eof = ^D; eol = <undef>;
> > eol2 = <undef>; swtch = <undef>; start = <undef>; stop = <undef>; susp = <undef>;
> > rprnt = ^R; werase = ^W; lnext = ^V; flush = ^O; min = 1; time = 0;
> > -parenb -parodd cs8 -hupcl -cstopb cread -clocal -crtscts
> > -ignbrk brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl ixon -ixoff
> > -iuclc -ixany -imaxbel -iutf8
> > -opost -olcuc -ocrnl -onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
> > isig -icanon -iexten -echo -echoe -echok -echonl -noflsh -xcase -tostop -echoprt
> > -echoctl -echoke
> > 
> > while this is running in a terminal:
> > 
> > speed 38400 baud; rows 40; columns 80; line = 0;
> > intr = ^C; quit = ^\; erase = ^H; kill = ^U; eof = ^D; eol = <undef>;
> > eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
> > werase = ^W; lnext = ^V; flush = ^O; min = 1; time = 0;
> > -parenb -parodd cs8 -hupcl -cstopb cread -clocal -crtscts
> > -ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
> > -iuclc -ixany -imaxbel -iutf8
> > opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
> > isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
> > echoctl echoke
> > 
> > You might notice that icrnl differs between the two cases.  There are other
> > differences (see below) but that detail might be relevant.
> > 
> > --- /tmp/piped	2016-07-24 14:58:27.986060176 -0400
> > +++ /tmp/normal	2016-07-24 14:58:45.842585062 -0400
> > @@ -1,4 +1,4 @@
> > -brkint
> > +-brkint
> >  bs0
> >  -clocal
> >  columns 80
> > @@ -7,11 +7,11 @@
> >  -crtscts
> >  cs8
> >  -cstopb
> > --echo
> > --echoctl
> > --echoe
> > --echok
> > --echoke
> > +echo
> > +echoctl
> > +echoe
> > +echok
> > +echoke
> >  -echonl
> >  -echoprt
> >  eof = ^D
> > @@ -21,9 +21,9 @@
> >  ff0
> >  flush = ^O
> >  -hupcl
> > --icanon
> > --icrnl
> > --iexten
> > +icanon
> > +icrnl
> > +iexten
> >  -ignbrk
> >  -igncr
> >  -ignpar
> > @@ -48,10 +48,10 @@
> >  -ofdel
> >  -ofill
> >  -olcuc
> > --onlcr
> > +onlcr
> >  -onlret
> >  -onocr
> > --opost
> > +opost
> >  -parenb
> >  -parmrk
> >  -parodd
> > @@ -59,9 +59,9 @@
> >  rows 40
> >  rprnt = ^R
> >  speed 38400 baud
> > -start = <undef>
> > -stop = <undef>
> > -susp = <undef>
> > +start = ^Q
> > +stop = ^S
> > +susp = ^Z
> >  swtch = <undef>
> >  tab0
> >  time = 0

So that's what I was fearing: no escape sequences. It must be done by accessing the terminal driver (TTY/PTY/...). In case of SSH, I don't know how it works at all. Is there a client-side terminal driver for that?

**Comment by cronvel** [comment]
@jwarkentin There is a native module that can help you: [termios](https://github.com/Gottox/node-termios). It can set up all those mode.

ATM, I'm don't know if it should be integrated directly to Terminal-Kit, or if the doc should simply link that module.

**Comment by jwarkentin** [comment]
Thanks for pointing me in the right direction. I didn't end up needing to use the termios module but it sent me down a sort of deep rabbit hole figuring out how all of this works. The difficulty of figuring out what to do was in the fact that there is basically no documentation anywhere that I could find explaining how all this was supposed to work.

When a user starts an SSH session that client side terminal is put into raw mode. The server they are connecting to is responsible for allocating a PTY to handle all of the input/output, mode settings and control code processing - all of which is performed by the kernel with a properly allocated PTY. So after I figured out how to do all of that (had to use the [pty.js](https://github.com/chjj/pty.js/) module) I had to pipe the SSH stream to the PTY master, which pushes to its slave (in `/dev/pts`) which I wrapped in a stream and then augmented a bit and passed into terminal-kit as the value for `stdin`, `stdout` and `stderr`.

So in short, I don't think it would make sense for terminal-kit to do anything with PTY's or termios. That really falls within the domain of the SSH connection and PTY/PTS file descriptor management. However, now that I've sorted all this out I no longer need the `options.crlf` option you previously added for me. I appreciate your help on all of this though.

**Comment by tomachinz** [comment]
Give some of these a try perhaps, these are Javascript / Node, at some point I realised stdout was better as it doesn't add a newline, some are better for linux versus macOS etc:
```
          console.log('\033c');
          console.log('\x1Bc');
          process.stdout.write('\x1B[2J\x1B[0f'); // CURSOR TO TOP LEFT????
          process.stdout.write("\x1B[2J"); // CLEAR TERMINAL SCREEN????
          process.stdout.write("\033[<0>;<0>H"); // pretty good
          process.stdout.write("\033[<0>;<0>f"); // cursor to 0,0
          process.stdout.write('\033c'); // <-- this is really the best one
```
